### ISSUE TITLE
Ansible is not creating proper directory structure in ~/.ansible

### ISSUE THREAD

**Issue by cadavre** [ansible/ansible#17876]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

ansible-playbook
##### ANSIBLE VERSION

```
ansible 2.1.2.0
  config file = 
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

Ubuntu 14.04

```
# ssh -V
OpenSSH_6.6.1p1 Ubuntu-2ubuntu2.8, OpenSSL 1.0.1f 6 Jan 2014
```
##### SUMMARY

There is missing `~/.ansible/cp` directory. While executing `ssh` with `ControlPath` OpenSSH – no directory inside on `~/.ansible/cp/` can be created because of missing parent dir.
##### STEPS TO REPRODUCE

```
Install Ansible 2.1.2.0 from apt or pip. Check ~/.ansible directory.
```

##### ACTUAL RESULTS

```
<myhost.com> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<jzapp.io> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<jzapp.io> SSH: ANSIBLE_REMOTE_PORT/remote_port/ansible_port set: (-o)(Port=22)
<jzapp.io> SSH: ANSIBLE_PRIVATE_KEY_FILE/private_key_file/ansible_ssh_private_key_file set: (-o)(IdentityFile="ansible/keys/jzappio-key.pem")
<jzapp.io> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<jzapp.io> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=ubuntu)
<jzapp.io> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<jzapp.io> SSH: PlayContext set ssh_common_args: ()
<jzapp.io> SSH: PlayContext set ssh_extra_args: ()
<jzapp.io> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r)
<jzapp.io> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=22 -o 'IdentityFile="ansible/keys/jzappio-key.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r jzapp.io '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /tmp/ansible-tmp-1475505073.67-7461893316535 `" && echo ansible-tmp-1475505073.67-7461893316535="` echo /tmp/ansible-tmp-1475505073.67-7461893316535 `" ) && sleep 0'"'"''
fatal: [jzappio]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
```

<myhost.com> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<myhost.com> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<myhost.com> SSH: ANSIBLE_REMOTE_PORT/remote_port/ansible_port set: (-o)(Port=22)
<myhost.com> SSH: ANSIBLE_PRIVATE_KEY_FILE/private_key_file/ansible_ssh_private_key_file set: (-o)(IdentityFile="ansible/keys/myhost-key.pem")
<myhost.com> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<myhost.com> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=ubuntu)
<myhost.com> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<myhost.com> SSH: PlayContext set ssh_common_args: ()
<myhost.com> SSH: PlayContext set ssh_extra_args: ()
<myhost.com> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r)
<myhost.com> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=22 -o 'IdentityFile="ansible/keys/myhost-key.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o ControlPath=/root/.ansible/cp/ansible-ssh-%h-%p-%r myhost.com '/bin/sh -c '"'"'( umask 77 && mkdir -p "`echo /tmp/ansible-tmp-1475505073.67-7461893316535`" && echo ansible-tmp-1475505073.67-7461893316535="`echo /tmp/ansible-tmp-1475505073.67-7461893316535`" ) && sleep 0'"'"''
fatal: [myhost]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}

```

While executing above `ssh` command in terminal:
```

...
debug1: setting up multiplex master socket
debug3: muxserver_listen: temporary control path /root/.ansible/cp/ansible-ssh-myhost.com-22-ubuntu.VMODcuk3ZKNVqT6S
muxserver_listen bind(): No such file or directory

```

I'm not sure which part is responsible for creation of `~/.ansible/cp` but it's definitely missing.
```

**Comment by abadger** [comment]
I'm unable to reproduce this.

```
$ pip install --user ansible
[...]
$ ~/.local/bin/ansible --version
ansible 2.1.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

$ rm -rf ~/.ansible
badger@ubuntu14:~$ ~/.local/bin/ansible -m ping -i '127.0.0.2,' all -c ssh -vvvv
Using /etc/ansible/ansible.cfg as config file
Loaded callback minimal of type stdout, v2.0
<127.0.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<127.0.0.2> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%h-%p-%r 127.0.0.2 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1477338818.85-163669845425631 `" && echo ansible-tmp-1477338818.85-163669845425631="` echo $HOME/.ansible/tmp/ansible-tmp-1477338818.85-163669845425631 `" ) && sleep 0'"'"''
<127.0.0.2> PUT /tmp/tmpcQ7Xdn TO /home/badger/.ansible/tmp/ansible-tmp-1477338818.85-163669845425631/ping
<127.0.0.2> SSH: EXEC sftp -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%h-%p-%r '[127.0.0.2]'
<127.0.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<127.0.0.2> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%h-%p-%r 127.0.0.2 '/bin/sh -c '"'"'chmod u+x /home/badger/.ansible/tmp/ansible-tmp-1477338818.85-163669845425631/ /home/badger/.ansible/tmp/ansible-tmp-1477338818.85-163669845425631/ping && sleep 0'"'"''
<127.0.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<127.0.0.2> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/badger/.ansible/cp/ansible-ssh-%h-%p-%r -tt 127.0.0.2 '/bin/sh -c '"'"'LANG=en_IN LC_ALL=en_IN LC_MESSAGES=en_IN /usr/bin/python /home/badger/.ansible/tmp/ansible-tmp-1477338818.85-163669845425631/ping; rm -rf "/home/badger/.ansible/tmp/ansible-tmp-1477338818.85-163669845425631/" > /dev/null 2>&1 && sleep 0'"'"''
127.0.0.2 | SUCCESS => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": null
        }, 
        "module_name": "ping"
    }, 
    "ping": "pong"
}
```

The code in question creates the directory and then verifies that the directory exists and is writable at the same time as the ControlPath is added: https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/ssh.py#L230

Looking closer at what you're doing, I notice that the controlpath error is coming about when you try to run ssh manually rather than when you try to run ansible.  So I don't know if creating the directory is the problem or if something else is interfering and then some cleanup happens which makes .ansible/cp go away before you try running ssh manually.
- What else can you tell us to help to reproduce this?
- What happens if you just ssh ubuntu@myhost.com?
- Do you have any ansible.cfg settings?
- If you manually create /root/.ansible/cp  does the ssh command you cut and pasted work?

**Comment by abadger** [comment]
needs_info

**Comment by ansibot** [comment]
@cadavre ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by cadavre** [comment]
I'm facing this issue when using Ansible inside of a Docker container. I'll share it with you and a use-case.

**Comment by jctanner** [comment]
@cadavre we're still waiting on the use-case.

needs_info

**Comment by ansibot** [comment]
@cadavre This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@cadavre This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@cadavre You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
